File: source/compiler/ast/Tuple.java
package compiler.ast;

import compiler.ir.Immediate;
import compiler.ir.Push;
import compiler.semantic.SemanticException;

import java.util.function.Function;

public class Tuple implements S {
    private S car, cdr;
    public Tuple(S car, S cdr) {
        this.car = car;
        this.cdr = cdr;
    }
    @Override
    public String toString() {
        return "(" + car.toString() + "." + cdr.toString() + ")";
    }
    @Override
    public Push evaluate() throws SemanticException {
        car.evaluate();
        cdr.evaluate();

        if (!(tmp0 instanceof Function<?, ?>)) {
            throw new SemanticException();
        }
        Function<?, ?> tmp2 = (Function<?, ?>) tmp0;
        R tmp4 = tmp2.apply(tmp1);
        return new Push(new Immediate<>(tmp4));
    }
    public T apply() {

    }
}

File: source/compiler/ast/Tuple.kt
package compiler.ast

import compiler.ir.Immediate
import compiler.ir.Push
import compiler.semantic.SemanticException
import java.util.function.Function

@Suppress("UNCHECKED_CAST")
class Tuple<T, R>(private val car: S, private val cdr: S) : S {
    override fun toString(): String {
        return "($car.$cdr)"
    }
    @Throws(SemanticException::class)
    override fun evaluate(): Push {
        val tmp0 = car.evaluate()
        val tmp1 = cdr.evaluate()
        if (!(tmp0 is Function<*, *> && tmp1 is T)) {
            throw SemanticException()
        }
        val tmp2 = tmp0 as Function<T, R>
        val tmp3 = tmp1 as T
        val tmp4 = tmp2.apply(tmp3)
        return Push(Immediate(tmp4))
    }
    fun apply(): T {

    }
}